mlir/lib/Bindings/Python/MainModule.cpp

```diff
@@ -34,7 +34,6 @@ PYBIND11_MODULE(_mlir, m) {
           "append_dialect_search_prefix",
           [](PyGlobals &self, std::string moduleName) {
             self.getDialectSearchPrefixes().push_back(std::move(moduleName));
-            self.clearImportCache();
```

Comment:
The only question is whether there's any value in preserving these semantics for the  - i.e., that when you append a dialect search prefix, that `llvm::StringSet<> loadedDialectModules` should be cleared thereby triggering "reloads" of existing dialect modules. 

I put "reload" in quotes because according to my understanding a reimport doesn't actually trigger a reload of the module[^1]. So even if someone is dynamically moving around existing modules (expecting dialect modules found under former prefixes to be found elsewhere post a call to `append_dialect_search_prefix`), the current implementation doesn't actually do that (since it reimports by calling `py::module::import(moduleName.c_str())`).

[^1]: You need to use `importlib` [to accomplish this](https://stackoverflow.com/questions/437589/how-do-i-unload-reload-a-python-module/437591#437591).

Reply:
My interpretation is that this is to support:
1. `loadDialectModule("my_dialect")` - doesn't find module
2. `append_dialect_search_prefix("path/to/my_dialect")`
3. `loadDialectModule("my_dialect")` - finds module
 
So its not about reloading existing modules, just loading new ones. Without clearing `loadedDialectModules` you'd end up hitting the early exit in `loadDialectModule`.

Reply:
Ah right that makes sense. Cool I'll restore the "clear" then.

Reply:
On second thought, let's make `loadDialectModule` do the right thing - report whether a dialect is actually registered or not (instead of just "pretending" a dialect is loaded).